service/sharddistributor/leader/process/processor.go

```diff
@@ -119,12 +126,152 @@ func (p *namespaceProcessor) runProcess(ctx context.Context) {
 	ticker := p.timeSource.NewTicker(p.cfg.Period)
 	defer ticker.Stop()
 
+	// Perform an initial rebalance on startup.
+	p.rebalanceShards(ctx)
+
+	// Subscribe to state changes from the store.
+	updateChan, err := p.shardStore.Subscribe(ctx)
+	if err != nil {
+		p.logger.Error("Failed to subscribe to state changes, stopping process.", tag.Error(err))
+		return
+	}
+
 	for {
 		select {
 		case <-ctx.Done():
 			p.logger.Info("Process cancelled")
 			return
+		case latestRevision, ok := <-updateChan:
```

Comment:
This will, I believe, trigger every time we get a heartbeat, which might be too much - I'd suggest adding a metric on how often rebalance is called

Reply:
Good point. I want to do it in a separate PR. 
Add metrics.Client, add metric definitions and metric emission.